src/core/lib/gprpp/sync.h

```diff
@@ -0,0 +1,126 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* Inline implementation of synchronization primitives, preferred over the C
+   public API for C++ code. */
+
+#ifndef GRPC_CORE_LIB_GPRPP_SYNC_H
+#define GRPC_CORE_LIB_GPRPP_SYNC_H
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/atm.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+
+#include <assert.h>
+#include <stdint.h>
+#include <atomic>
+#include <type_traits>
+
+#include "src/core/lib/gprpp/memory.h"
+
+namespace grpc_core {
+
+class RefCount {
+ public:
+  explicit RefCount(intptr_t init = 1) : val(init) {}
+
+  // Reinitializes the ref-count.
+  void Reset(intptr_t init) { val.store(init, std::memory_order_relaxed); }
```

Comment:
What's the use-case for this?

Reply:
Here is the main user:
https://github.com/grpc/grpc/blob/b41a44dc2e001358602e8e885c674ddc49387edd/src/core/lib/surface/call.cc#L1851

Reply:
If that's the main use-case, is there a way we could restructure the code there such that we can just instantiate a new `RefCount` whenever we start a batch?  For example, when we re-use an existing `batch_control` struct, we can use placement-new to instantiate it, and then call its dtor when it gets unreffed.  Then maybe we can change `call_start_batch()` to do two passes over the passed-in ops, so that we know how many refs we need before we instantiate the `batch_control` struct.  Or, if that two-pass approach is too expensive, maybe we can instead use `grpc_core::ManualConstructor<>`, so that we can instantiate the `RefCount` after the `batch_control` is already instantiated.

If there is no alternative that doesn't come with performance penalties, I can live with this.  But having an explicit reset method feels fairly ugly to me from an API perspective, since the only time it should ever be used is when the previous refcount has already gone to zero and we are now re-using the same object.  So I'd prefer to avoid it if possible.